.github/workflows/ci.yml

```diff
@@ -0,0 +1,327 @@
+        run: |
+          PREFIX=${HOME}/.unit
+          BIN=${PREFIX}/bin
+          VAR=${PREFIX}/var
+          mkdir -p $BIN
+          mkdir -p $VAR
+
+          echo "prefix=${PREFIX}" >> "$GITHUB_OUTPUT"
+          echo "bin=${BIN}" >> "$GITHUB_OUTPUT"
+          echo "bin=${BIN}" >> "$GITHUB_PATH"
+          echo "var=${VAR}" >> "$GITHUB_OUTPUT"
+          cat "$GITHUB_OUTPUT"
+
+      # Provides module, language version and testpath from build name
+      - name: Output build metadata
+        id: metadata
+        run: |
+          # Split the build name by '-' into module and version
+          IFS='-' read -r module version <<< "${{ matrix.build }}"
+
+          testpath="test/test_${module}*"
+
+          # Run all tests for "unit" and "python"
+          # Python is the default module for tests
+          if [ "$module" = "unit" ] || [ "$module" = "python" ]; then
+            testpath="test"
+          fi
+
+          echo "module=${module}" >> "$GITHUB_OUTPUT"
+          echo "version=${version}" >> "$GITHUB_OUTPUT"
+          echo "testpath=${testpath}" >> "$GITHUB_OUTPUT"
+
+          NJS_VERSION=$(sed -n "s/NJS_VERSION := \(.*\)/\1/p" pkg/contrib/src/njs/version)
+          echo "njs_version=${NJS_VERSION}" >> "$GITHUB_OUTPUT"
+
+          cat "$GITHUB_OUTPUT"
+
+      # https://github.com/actions/runner-images/issues/2821
+      - name: Kill mono process
+        run: |
+          sudo systemctl stop mono-xsp4.service
+          sudo systemctl mask mono-xsp4.service
+          sudo systemctl status mono-xsp4.service || true
+          PID=$(sudo lsof -t -i :8084)
+          echo "Killing PID $PID"
+          sudo kill -9 $PID
+
+      ##
+      ## njs
+      ##
+
+      - name: Clone njs repository
+        uses: actions/checkout@v4
+        with:
+          repository: nginx/njs
+          ref: '${{ steps.metadata.outputs.njs_version }}'
+          path: njs
+
+      - name: Make njs
+        run: |
+          ./configure --no-libxml2 --no-zlib
+          make -j4 -k
+        working-directory: njs
+
+      ##
+      ## Unit
+      ##
+
+      - name: Configure unit
+        run: |
+          ./configure \
+            --prefix=${{ steps.dir.outputs.prefix }} \
+            --sbindir=${{ steps.dir.outputs.bin }} \
+            --logdir=${{ steps.dir.outputs.var }}/log \
+            --log=${{ steps.dir.outputs.var }}/log/unit/unit.log \
+            --runstatedir=${{ steps.dir.outputs.var }}/run \
+            --pid=${{ steps.dir.outputs.var }}/run/unit/unit.pid \
+            --control=unix:${{ steps.dir.outputs.var }}/run/unit/control.sock \
+            --modules=${{ steps.dir.outputs.prefix }}/lib/unit/modules \
+            --statedir=${{ steps.dir.outputs.var }}/state/unit \
+            --tests \
+            --openssl \
+            --njs \
+            --cc-opt="-I njs/src/ -I njs/build"  \
+            --ld-opt="-L njs/build" \
+            --debug
+
+      - name: Make unit
+        run: |
+          make -j4 -k
```

Comment:
You probably want `make -j4 -k || make`.  Otherwise, the error output will be ugly.  Cc: @thresheek

Reply:
Hmm, from doing some simple testing

```
$ make -j4 -k || make
```
isn't much different from just
```
$ make -j4
```

They both clearly show the first error.

With the `-k` option you do have the opportunity to go through the scrollback to see other errors, but with the second make you'll only see the first error... (unless any other errors just happen to be in the last compiled files).

But in GH,  you're probably going to have to scroll through the compiler output anyway to get to the bottom, so I'm not really sure the second make buys us much...

Reply:
> Hmm, from doing some simple testing
> 
> ```
> $ make -j4 -k || make
> ```
> 
> isn't much different from just
> 
> ```
> $ make -j4
> ```
> 
> They both clearly show the first error.

No, they don't:

```mk
alx@debian:~/tmp$ cat Makefile 
.PHONY: all a b
all: a b

a:
	$(info	A)
	sleep 1
	echo errors from a
	false

b:
	$(info	B)
	echo errors from b
	false
```

Simple make:
```sh
alx@debian:~/tmp$ make -k
A
sleep 1
echo errors from a
errors from a
false
make: *** [Makefile:8: a] Error 1
B
echo errors from b
errors from b
false
make: *** [Makefile:13: b] Error 1
make: Target 'all' not remade because of errors.
```

Ugly, parallel make:
```
alx@debian:~/tmp$ make -j4 -k
A
sleep 1
B
echo errors from b
errors from b
false
make: *** [Makefile:13: b] Error 1
echo errors from a
errors from a
false
make: *** [Makefile:8: a] Error 1
make: Target 'all' not remade because of errors.
```

Ugly parallel make, followed by simple make:
```
alx@debian:~/tmp$ make -j4 -k || make
A
sleep 1
B
echo errors from b
errors from b
false
make: *** [Makefile:13: b] Error 1
echo errors from a
errors from a
false
make: *** [Makefile:8: a] Error 1
make: Target 'all' not remade because of errors.
A
sleep 1
echo errors from a
errors from a
false
make: *** [Makefile:8: a] Error 1
```
In the last case, you'll see the initial part is mixed, due to parallel make, but you can just ignore it, and start reading the last part.

See <https://github.com/shadow-maint/shadow/commit/97f79e3b2715d78c9342cbb292e1a3e72a23fe62>

> 
> With the `-k` option you do have the opportunity to go through the scrollback to see other errors, but with the second make you'll only see the first error... (unless any other errors just happen to be in the last compiled files).

> 
> But in GH, you're probably going to have to scroll through the compiler output anyway to get to the bottom, so I'm not really sure the second make buys us much...

Reply:
I agree with @alejandro-colomar - the '-k || make' trick is really helpful with multiple build targets' outputs are intertwined.   We've been using it in buildbot builds for a few years too.

Reply:
> No, they don't:

Oh, but I'm afraid they do...

```
$ make -j4 -k || make
...
src/nxt_file.c: In function ‘nxt_file_chown’:
src/nxt_file.c:357:52: error: ‘dummy’ undeclared (first use in this function)
  357 |                                  owner, nxt_errno, dummy,
      |                                                    ^~~~~
src/nxt_log.h:81:46: note: in definition of macro ‘nxt_log_alert’
   81 |         _log_->handler(NXT_LOG_ALERT, _log_, __VA_ARGS__);                    \
      |                                              ^~~~~~~~~~~
src/nxt_file.c:356:13: note: in expansion of macro ‘nxt_thread_log_alert’
  356 |             nxt_thread_log_alert("getpwnam_r(\"%s\", ...) failed %E %s",
      |             ^~~~~~~~~~~~~~~~~~~~
src/nxt_file.c:357:52: note: each undeclared identifier is reported only once for each function it appears in
  357 |                                  owner, nxt_errno, dummy,
      |                                                    ^~~~~
src/nxt_log.h:81:46: note: in definition of macro ‘nxt_log_alert’
   81 |         _log_->handler(NXT_LOG_ALERT, _log_, __VA_ARGS__);                    \
      |                                              ^~~~~~~~~~~
src/nxt_file.c:356:13: note: in expansion of macro ‘nxt_thread_log_alert’
  356 |             nxt_thread_log_alert("getpwnam_r(\"%s\", ...) failed %E %s",
      |             ^~~~~~~~~~~~~~~~~~~~
make: *** [build/Makefile:226: build/src/nxt_file.o] Error 1
$
```

```
$ make -j4
...
src/nxt_file.c: In function ‘nxt_file_chown’:
src/nxt_file.c:357:52: error: ‘dummy’ undeclared (first use in this function)
  357 |                                  owner, nxt_errno, dummy,
      |                                                    ^~~~~
src/nxt_log.h:81:46: note: in definition of macro ‘nxt_log_alert’
   81 |         _log_->handler(NXT_LOG_ALERT, _log_, __VA_ARGS__);                    \
      |                                              ^~~~~~~~~~~
src/nxt_file.c:356:13: note: in expansion of macro ‘nxt_thread_log_alert’
  356 |             nxt_thread_log_alert("getpwnam_r(\"%s\", ...) failed %E %s",
      |             ^~~~~~~~~~~~~~~~~~~~
src/nxt_file.c:357:52: note: each undeclared identifier is reported only once for each function it appears in
  357 |                                  owner, nxt_errno, dummy,
      |                                                    ^~~~~
src/nxt_log.h:81:46: note: in definition of macro ‘nxt_log_alert’
   81 |         _log_->handler(NXT_LOG_ALERT, _log_, __VA_ARGS__);                    \
      |                                              ^~~~~~~~~~~
src/nxt_file.c:356:13: note: in expansion of macro ‘nxt_thread_log_alert’
  356 |             nxt_thread_log_alert("getpwnam_r(\"%s\", ...) failed %E %s",
      |             ^~~~~~~~~~~~~~~~~~~~
make: *** [build/Makefile:226: build/src/nxt_file.o] Error 1
make: *** Waiting for unfinished jobs....
$
```

Reply:
> See [shadow-maint/shadow@97f79e3](https://github.com/shadow-maint/shadow/commit/97f79e3b2715d78c9342cbb292e1a3e72a23fe62)

Hmm, you did it slightly differently there and not that Iker was really convinced either way...

Reply:
> > See [shadow-maint/shadow@97f79e3](https://github.com/shadow-maint/shadow/commit/97f79e3b2715d78c9342cbb292e1a3e72a23fe62)
> 
> Hmm, you did it slightly differently there and not that Iker was really convinced either way...

Yeah, take your favourite one.  Each one has advantages and disadvantages.  I suggested `|| make` because that's what BB is doing already, as K said.

> > No, they don't:
> 
> Oh, but I'm afraid they do...
> 

What you're seeing is racy, and the race just happens to behave relatively well in your case.  In your case, probably just the command is separate from the error, which is bad, but not the worst thing that can happen  Output from different compilations may get mixed very badly (I've seen it in other projects).

See:

```
$ MANWIDTH=72 man make | sed -n '/^ *-O/,/^$/p'
     -O[type], --output-sync[=type]
          When  running  multiple jobs in parallel with ‐j, ensure the
          output of each job is collected together rather than  inter‐
          spersed  with output from other jobs.  If type is not speci‐
          fied or is target the output from the entire recipe for each
          target is grouped together.  If type is line the output from
          each command line within a recipe is grouped  together.   If
          type  is  recurse  output  from  an entire recursive make is
          grouped together.  If type is none output synchronization is
          disabled.

```

From that documentation it's unclear what the default is, but it's `none`.  You can check GNU Make's source code:

```c
$ grepc decode_output_sync_flags src/
src/main.c:static void
decode_output_sync_flags (void)
{
#ifdef NO_OUTPUT_SYNC
  output_sync = OUTPUT_SYNC_NONE;
#else
  if (output_sync_option)
    {
      if (streq (output_sync_option, "none"))
        output_sync = OUTPUT_SYNC_NONE;
      else if (streq (output_sync_option, "line"))
        output_sync = OUTPUT_SYNC_LINE;
      else if (streq (output_sync_option, "target"))
        output_sync = OUTPUT_SYNC_TARGET;
      else if (streq (output_sync_option, "recurse"))
        output_sync = OUTPUT_SYNC_RECURSE;
      else
        OS (fatal, NILF,
            _("unknown output-sync type '%s'"), output_sync_option);
    }

  if (sync_mutex)
    osync_parse_mutex (sync_mutex);
#endif
}
```

Reply:
Alternatively, you may want to just specify `-Otarget`

Reply:
I'm happy to use `make -j4 -k || make` for the time being to match what we're doing elsewhere. We can do improvements in future PRs for those who are passionate about improving our error visibility.

Again, this is just an interim / first-pass at an Actions build for Unit alone, and I suspect much will change with time 😄

Reply:
@ac000 

Today, I've seen a mixed line in a multi-process make, while testing some improvement to the Linux man-pages' build system (I'm adding a `distcheck` target).

Here's a line that was printed:
```
MAKE build-pdf (2>/dev/null):MAKE build-ps (-i): GROPS	.tmp/man/man7/koi8-r.7.ps
 GROPDF	.tmp/man/man4/sd.4.pdf
```
while it should have been
```
MAKE build-pdf (2>/dev/null): GROPDF	.tmp/man/man4/sd.4.pdf
MAKE build-ps (-i): GROPS	.tmp/man/man7/koi8-r.7.ps
```

The command was
```sh
$ make distcheck TMPDIR1=/tmp/tmp1 TMPDIR2=/tmp/tmp2 -j24 -k
```
You can guess that 24 processes competing for /dev/stdout don't behave well.

And the relevant makefile code is:
```mk
$(builddir)/distcheck.build-ps.touch: $(builddir)/distcheck.build-pre.touch
$(builddir)/distcheck.build-ps.touch: $(TMPDIR1)/$(DISTNAME) | $$(@D)/
	$(info	MAKE		build-ps)
	$(MAKE) -C $< -k build-ps $(HIDE_ERR) \
	| $(TS) 'MAKE build-ps (2>/dev/null):' \
	|| $(TRUE)
	$(MAKE) -C $< -i nothing $(DISTCHECK_IGNORE_BUILD_PS) $(HIDE_ERR) \
	| $(TS) 'MAKE build-ps (-i):'
	$(MAKE) -C $< build-ps \
	| $(TS) 'MAKE build-ps:'
	$(TOUCH) $@

$(builddir)/distcheck.build-pdf.touch: $(builddir)/distcheck.build-pre.touch
$(builddir)/distcheck.build-pdf.touch: $(TMPDIR1)/$(DISTNAME) | $$(@D)/
	$(info	MAKE		build-pdf)
	$(MAKE) -C $< -k build-pdf $(HIDE_ERR) \
	| $(TS) 'MAKE build-pdf (2>/dev/null):' \
	|| $(TRUE)
	$(MAKE) -C $< -i nothing $(DISTCHECK_IGNORE_BUILD_PDF) $(HIDE_ERR) \
	| $(TS) 'MAKE build-pdf (-i):'
	$(MAKE) -C $< build-pdf \
	| $(TS) 'MAKE build-pdf:'
	$(TOUCH) $@
```

Reply:
We've merged the patch with `-Orecurse` in shadow:

-  <https://github.com/shadow-maint/shadow/pull/950>

You may be interested in doing something similar here.